hudi-flink/src/main/java/org/apache/hudi/table/HoodieTableSource.java

```diff
@@ -312,14 +318,18 @@ private DataType getProducedDataType() {
     TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
     final Schema tableAvroSchema;
     try {
-      tableAvroSchema = schemaUtil.getTableAvroSchema();
+      // always includes the change flag
+      tableAvroSchema = schemaUtil.getTableAvroSchema(true);
     } catch (Exception e) {
       throw new HoodieException("Get table avro schema error", e);
     }
     final DataType rowDataType = AvroSchemaConverter.convertToDataType(tableAvroSchema);
     final RowType rowType = (RowType) rowDataType.getLogicalType();
     final RowType requiredRowType = (RowType) getProducedDataType().notNull().getLogicalType();
 
+    final Schema.Field operationField = tableAvroSchema.getField(HoodieRecord.OPERATION_METADATA_FIELD);
+    final int operationPos = operationField == null ? -1 : operationField.pos();
```

Comment:
We can move the `operationPos` into the `MergeOnReadTableState` to simplify the logic.